File: library-window/src/wuyi_seat_bot/server_sync/sync_planner.py
"""Manual_Sync_Action 差异规划器。

本模块只承担「按 ``student_id`` 对比服务端活跃池快照与本地账号集合的受管字段
差异」这一项纯函数职责，与 design.md「数据流 4：Manual_Sync_Action 全流程」
中给出的 diff 算法等价。

设计依据：

- Requirement 13.5 / 13.10：Manual_Sync_Action 在拉取接口 A / 接口 B 后，弹
  Sync_Coverage_Confirmation 时按账号 (``student_id``) 行级展示「新增 / 替换 /
  移除」候选条目。
- Requirement 13.14：差异计算结果 ``SHALL`` 不因 Sync_Selection 的勾选变化而
  改变；Sync_Selection 仅作用于「应用差异」阶段。本模块只做差异计算，不维
  护任何 Sync_Selection 状态。
- Requirement 13.17 (旧 13-Q1 默认)：Manual_Sync_Action 是「按学号 diff」而非
  「全量覆盖」；本模块严格按 ``student_id`` 做集合比对，不做模糊匹配。

调用约束：

- 本模块 **不** 修改 Local_Account_Store；只产出 :class:`SyncCandidate`
  列表，由 ``sync_applier``（任务 11.11）按用户勾选状态执行覆盖。
- 本模块 **不** 依赖运行时网络状态；输入 ``server_snapshot`` 与 ``local_store``
  均由调用方在外层准备好。

受管字段集合（Requirement 13.6 / 13.13、design「Sync_Coverage_Confirmation
数据结构与字段口径」）::

    {student_id, password, display_name, automation_tasks, custom_windows}

其中 ``custom_windows`` 是 ``automation_tasks[i].custom_windows`` 的递归项；
本模块在比较 ``automation_tasks`` 时一并对每个任务的 ``custom_windows``
做内容比较，不再单独建模顶层 ``custom_windows`` 字段。
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, Literal

from wuyi_seat_bot.server_sync.active_pool_repository import (
    ActiveAccountDetail,
    AutomationTask,
    CustomWindow,
)


# --------------------------------------------------------------------------- #
# 受管字段常量                                                                  #
# --------------------------------------------------------------------------- #

MANAGED_FIELDS: frozenset[str] = frozenset(
    {
        "student_id",
        "password",
        "display_name",
        "automation_tasks",
        "custom_windows",
    }
)
"""Manual_Sync_Action 覆盖时受服务端管理的字段集合。

用于文档与外部模块（``sync_applier`` / UI）共享口径，本模块在对比时按字段
名逐一展开比较；``custom_windows`` 通过 ``automation_tasks`` 内部嵌套递归
覆盖，不再单独建模顶层字段。
"""


# --------------------------------------------------------------------------- #
# Data classes                                                                #
# --------------------------------------------------------------------------- #


@dataclass(frozen=True, slots=True)
class LocalAutomationTask:
    """sync_planner 视角下的本地自动任务快照。

    与服务端 :class:`AutomationTask` 字段一致，但不携带 ``updated_at`` 与
    ``revision`` 等服务端元数据，便于做受管字段差异对比。
    """

    task_id: int
    room_name: str
    seat_number: str
    mode: str
    custom_windows: tuple[CustomWindow, ...]
    enabled: bool


@dataclass(frozen=True, slots=True)
class LocalAccountSummary:
    """sync_planner 视角下的本地账号快照。

    Attributes:
        account_name: 本地配置中的账号名；用于把同步结果落回对应账号。
        student_id: 学号；与服务端清单的 ``student_id`` 字段对齐。
        password: 本地保存的密码（明文）。
        display_name: 备注 / 姓名。
        automation_tasks: 本地自动任务集合，顺序无关（比较时会规范化排序）。
        is_server_managed: 是否被「受服务端管理活跃账号」标记。仅当为
            ``True`` 且服务端清单不再包含该 ``student_id`` 时，diff 算法
            才产出 ``kind == "remove"`` 的候选条目（Requirement 13.18）。
    """

    student_id: str
    password: str
    display_name: str
    automation_tasks: tuple[LocalAutomationTask, ...]
    is_server_managed: bool
    account_name: str = ""


@dataclass(frozen=True, slots=True)
class SyncCandidate:
    """Sync_Coverage_Confirmation 弹窗中的单条候选。

    Attributes:
        kind: 同步动作类型；``add`` / ``replace`` / ``remove``。
        student_id: 候选行对应的学号。
        server_payload: 服务端原始 ``ActiveAccountDetail``（含密码、自动任务
            等完整信息）。``add`` / ``replace`` 时必填，``remove`` 时为
            ``None``。
        local_summary: 本地受管字段快照。``replace`` / ``remove`` 时必填，
            ``add`` 时为 ``None``。
        default_checked: 弹窗首次打开时该行的默认勾选状态。
            ``add`` / ``replace`` 默认 ``True``；``remove`` 默认 ``False``
            （Requirement 13.11）。
    """

    kind: Literal["add", "replace", "remove"]
    student_id: str
    server_payload: ActiveAccountDetail | None
    local_summary: LocalAccountSummary | None
    default_checked: bool


# --------------------------------------------------------------------------- #
# 公共 API                                                                     #
# --------------------------------------------------------------------------- #


def compute_diff(
    server_snapshot: Iterable[ActiveAccountDetail],
    local_store: Iterable[LocalAccountSummary],
) -> list[SyncCandidate]:
    """按 ``student_id`` 计算服务端活跃池快照与本地账号集合的受管字段差异。

    与 design.md「数据流 4：Manual_Sync_Action 全流程」中 diff 算法等价：

    - **新增**：服务端有该 ``student_id``，本地无。
    - **替换**：两端 ``student_id`` 匹配，但任一受管字段（学号 / 密码 /
      备注 / 自动任务 / 自定义窗口）不一致。
    - **移除**：本地标记 ``is_server_managed=True`` 但服务端清单已无对应
      ``student_id``。

    Args:
        server_snapshot: 接口 A / 接口 B 返回的活跃池账号详情集合。可空。
        local_store: 本地账号集合的受管字段快照。可空。

    Returns:
        候选条目列表，按 (kind, student_id) 升序稳定排序，方便 UI 展示
        与单元测试断言。``add`` / ``replace`` 默认 ``default_checked=True``；
        ``remove`` 默认 ``default_checked=False``。

    Raises:
        ValueError: 若 ``server_snapshot`` 中存在重复 ``student_id``，或
            ``local_store`` 中存在重复 ``student_id``。同一学号在同一端只
            应出现一次（Requirement 1-Q3 唯一键约束）。
    """

    server_by_sid = _index_by_student_id(
        server_snapshot,
        get_sid=lambda detail: detail.student_id,
        side="server_snapshot",
    )
    local_by_sid = _index_by_student_id(
        local_store,
        get_sid=lambda summary: summary.student_id,
        side="local_store",
    )

    candidates: list[SyncCandidate] = []

    # 新增：服务端有、本地无
    for sid, srv in server_by_sid.items():
        if sid not in local_by_sid:
            candidates.append(
                SyncCandidate(
                    kind="add",
                    student_id=sid,
                    server_payload=srv,
                    local_summary=None,
                    default_checked=True,
                )
            )

    # 替换：两端 student_id 匹配，但任一受管字段不一致
    for sid, srv in server_by_sid.items():
        if sid in local_by_sid:
            local_summary = local_by_sid[sid]
            if not _managed_fields_equal(srv, local_summary):
                candidates.append(
                    SyncCandidate(
                        kind="replace",
                        student_id=sid,
                        server_payload=srv,
                        local_summary=local_summary,
                        default_checked=True,
                    )
                )

    # 移除：本地标记「受服务端管理」但服务端清单已无对应 student_id
    for sid, local_summary in local_by_sid.items():
        if local_summary.is_server_managed and sid not in server_by_sid:
            candidates.append(
                SyncCandidate(
                    kind="remove",
                    student_id=sid,
                    server_payload=None,
                    local_summary=local_summary,
                    default_checked=False,
                )
            )

    # 稳定排序，便于 UI 展示与测试断言。
    _KIND_ORDER = {"add": 0, "replace": 1, "remove": 2}
    candidates.sort(key=lambda c: (_KIND_ORDER[c.kind], c.student_id))
    return candidates


# --------------------------------------------------------------------------- #
# 内部辅助                                                                      #
# --------------------------------------------------------------------------- #


def _index_by_student_id(items, *, get_sid, side: str):
    """按 ``student_id`` 建立索引，重复学号抛 ValueError。"""

    indexed: dict[str, object] = {}
    for item in items:
        sid = get_sid(item)
        if not isinstance(sid, str) or not sid:
            raise ValueError(f"{side} 条目的 student_id 必须是非空字符串")
        if sid in indexed:
            raise ValueError(f"{side} 中存在重复 student_id：{sid!r}")
        indexed[sid] = item
    return indexed  # type: ignore[return-value]


def _managed_fields_equal(
    server: ActiveAccountDetail, local: LocalAccountSummary
) -> bool:
    """比较受管字段集合是否完全一致。

    顺序无关地比较 ``automation_tasks``，并对每个任务的 ``custom_windows``
    做内容比较（忽略服务端附加的 ``updated_at`` / ``revision`` 元数据）。
    """

    if server.student_id != local.student_id:
        return False
    if server.password != local.password:
        return False
    if server.display_name != local.display_name:
        return False
    if _normalize_server_tasks(server.automation_tasks) != _normalize_local_tasks(
        local.automation_tasks
    ):
        return False
    return True


def _normalize_server_tasks(
    tasks: Iterable[AutomationTask],
) -> tuple[LocalAutomationTask, ...]:
    """把服务端 :class:`AutomationTask` 序列归一化为受管字段元组并稳定排序。"""

    normalized = [
        LocalAutomationTask(
            task_id=t.task_id,
            room_name=t.room_name,
            seat_number=t.seat_number,
            mode=t.mode,
            custom_windows=_normalize_custom_windows(t.custom_windows),
            enabled=t.enabled,
        )
        for t in tasks
    ]
    return tuple(sorted(normalized, key=_task_sort_key))


def _normalize_local_tasks(
    tasks: Iterable[LocalAutomationTask],
) -> tuple[LocalAutomationTask, ...]:
    """把本地 :class:`LocalAutomationTask` 序列做稳定排序，并归一化窗口序列。"""

    normalized = [
        LocalAutomationTask(
            task_id=t.task_id,
            room_name=t.room_name,
            seat_number=t.seat_number,
            mode=t.mode,
            custom_windows=_normalize_custom_windows(t.custom_windows),
            enabled=t.enabled,
        )
        for t in tasks
    ]
    return tuple(sorted(normalized, key=_task_sort_key))


def _normalize_custom_windows(
    windows: Iterable[CustomWindow],
) -> tuple[CustomWindow, ...]:
    """对自定义窗口序列做稳定排序，便于顺序无关比较。"""

    return tuple(
        sorted(
            windows,
            key=lambda w: (w.date, w.start_hour, w.end_hour),
        )
    )


def _task_sort_key(task: LocalAutomationTask) -> tuple:
    """自动任务排序键。

    优先按 ``task_id`` 排序；当多个任务共用同一 ``task_id``（理论上不应发生，
    但本模块不主动校验）时，按其它字段做次级排序，保证排序稳定。
    """

    return (
        task.task_id,
        task.room_name,
        task.seat_number,
        task.mode,
        task.enabled,
        task.custom_windows,
    )


__all__ = [
    "LocalAccountSummary",
    "LocalAutomationTask",
    "MANAGED_FIELDS",
    "SyncCandidate",
    "compute_diff",
]

File: library-android/app/src/main/java/com/wuyi/libraryauto/sync/SyncPlanner.kt
package com.wuyi.libraryauto.sync

/**
 * 任务 12.11：Manual_Sync_Action 候选条目计算器（与 Window 端 `sync_planner.py` 等价）。
 *
 * 设计依据：
 * - design.md「数据流 4：Manual_Sync_Action 全流程」中按 `student_id` 计算 add / replace /
 *   remove 三类候选条目的 diff 算法。
 * - design.md「Sync_Coverage_Confirmation 数据结构与字段口径」中 [SyncCandidate] 的字段
 *   契约（kind / student_id / server_payload / local_summary / default_checked）。
 * - Requirements 13.5 / 13.10 / 13.14 / 13.17：服务端清单 → 本地差异 → 候选条目。
 *
 * 行为契约：
 * - **不**修改 Room、不写入本地任何持久化层；纯函数式计算。
 * - **不**依赖运行时网络状态；输入是已经拉到的服务端快照与本地存储集合。
 * - 受管字段集合（与 Window 端、Android 端一致）：
 *   `student_id` / `password` / `display_name` / 关联 `automation_tasks` / `custom_windows`。
 * - 三类候选条目的判定口径：
 *   - `add`：服务端 `student_id` 命中、本地无 → `default_checked = true`。
 *   - `replace`：两端 `student_id` 匹配，但任一受管字段不一致 → `default_checked = true`。
 *   - `remove`：本地 [LocalAccount.isServerManaged] 标记为 true、但服务端清单已无该
 *     `student_id` → `default_checked = false`。
 * - 输出顺序：先按 `add → replace → remove` 分组（与 design 伪代码一致），同组内按
 *   `student_id` 字典序稳定排序，便于 UI 渲染时不抖动、便于属性测试断言确定性。
 */
object SyncPlanner {
    /**
     * 计算服务端快照与本地存储之间的同步差异，输出候选条目列表。
     *
     * @param serverSnapshot 接口 A + 接口 B 拉取后合并得到的活跃账号详情快照集合。
     *   每条要包含 `student_id` / `password` / `display_name` / 关联 `automation_tasks` /
     *   `custom_windows`，服务端口径为权威。
     * @param localStore 本地 Room 表中的当前账号集合。每条要包含同样的受管字段，且要带
     *   [LocalAccount.isServerManaged] 标记用于判定 remove 候选。
     * @return 候选条目列表。`add` / `replace` / `remove` 顺序与 design.md 伪代码一致；
     *   同组内按 `student_id` 字典序排序。
     */
    fun computeDiff(
        serverSnapshot: List<ServerAccountSnapshot>,
        localStore: List<LocalAccount>,
    ): List<SyncCandidate> {
        // design.md 伪代码：把两端按 student_id 索引，分三轮扫描产出候选条目。
        // 服务端快照若出现重复 student_id，以后出现者覆盖前者；本地存储同理。这一行为与
        // Window 端 `dict[..]=` 语义一致；上层应通过乐观假设保证唯一性，本计算器不抛错、
        // 不输出诊断告警，避免污染 UI 流程。
        val serverBySid: Map<String, ServerAccountSnapshot> =
            serverSnapshot.associateBy(ServerAccountSnapshot::studentId)
        val localBySid: Map<String, LocalAccount> =
            localStore.associateBy(LocalAccount::studentId)

        val addCandidates = mutableListOf<SyncCandidate>()
        val replaceCandidates = mutableListOf<SyncCandidate>()
        val removeCandidates = mutableListOf<SyncCandidate>()

        // 1) 新增：服务端有、本地无。
        for ((sid, server) in serverBySid) {
            if (sid !in localBySid) {
                addCandidates.add(
                    SyncCandidate(
                        kind = SyncKind.Add,
                        studentId = sid,
                        serverPayload = server,
                        localSummary = null,
                        defaultChecked = true,
                    ),
                )
            }
        }

        // 2) 替换：两端匹配，但任一受管字段不一致。
        for ((sid, server) in serverBySid) {
            val local = localBySid[sid] ?: continue
            if (hasManagedFieldDiff(server = server, local = local)) {
                replaceCandidates.add(
                    SyncCandidate(
                        kind = SyncKind.Replace,
                        studentId = sid,
                        serverPayload = server,
                        localSummary = local.toSummary(),
                        defaultChecked = true,
                    ),
                )
            }
        }

        // 3) 移除：本地受服务端管理、但服务端清单已无对应 student_id。
        for ((sid, local) in localBySid) {
            if (local.isServerManaged && sid !in serverBySid) {
                removeCandidates.add(
                    SyncCandidate(
                        kind = SyncKind.Remove,
                        studentId = sid,
                        serverPayload = null,
                        localSummary = local.toSummary(),
                        defaultChecked = false,
                    ),
                )
            }
        }

        // 同组内按 student_id 字典序稳定排序，输出顺序：add → replace → remove。
        val byStudentId = compareBy<SyncCandidate> { it.studentId }
        return addCandidates.sortedWith(byStudentId) +
            replaceCandidates.sortedWith(byStudentId) +
            removeCandidates.sortedWith(byStudentId)
    }

    /**
     * 受管字段差异判定。任一字段不一致即视为「需要替换」。
     *
     * 字段比较语义：
     * - `student_id`：两端按 student_id 匹配，理论上必相等；这里冗余比较保持算法对称。
     * - `password`：明文比较。空串视为「服务端密码未知」，但 design.md 接口 B 必返非空，
     *   不做特殊处理。
     * - `display_name`：直接字符串比较。空串与非空串语义不同。
     * - `automation_tasks`：只比较 Android 本地能持久化、能影响调度行为的字段子集，
     *   见 [hasAutomationTaskDiff]。
     */
    private fun hasManagedFieldDiff(
        server: ServerAccountSnapshot,
        local: LocalAccount,
    ): Boolean {
        if (server.studentId != local.studentId) return true
        if (server.password != local.password) return true
        if (server.displayName != local.displayName) return true
        if (hasAutomationTaskDiff(server.automationTasks, local.automationTasks)) return true
        return false
    }

    /**
     * 自动任务列表差异判定（按 [AutomationTaskDto.taskId] 主键比较）。
     *
     * 比较口径只覆盖 Android 本地 `automation_plans` schema 实际能落库、且能改变调度
     * 行为的字段：`taskId / roomName / seatNumber / customWindows / enabled`。`mode`
     * （服务端 preferred / manual / random）、`revision`（服务端乐观并发版本号）、
     * `updatedAt`（服务端写入时间戳）这些是服务端独有元数据，本地不持久化也不会基于
     * 它们调度，所以即便服务端响应中带了新 revision，也不应反复触发 replace 候选。
     *
     * 这样做的代价是：服务端把 mode 从 preferred 改成 manual、但其余字段都没动，本地
     * 不会主动提示用户去同步；可以通过 design.md 中其它显式触发点（比如用户在
     * Sync_Coverage_Confirmation 主动点「全选」）来覆盖。考虑到 Android 本地连
     * mode 都没法持久化，反复提示 replace 而又写不进去才是真正的死循环。
     */
    private fun hasAutomationTaskDiff(
        serverTasks: List<AutomationTaskDto>,
        localTasks: List<AutomationTaskDto>,
    ): Boolean {
        if (serverTasks.size != localTasks.size) return true
        val serverByTaskId = serverTasks.associateBy(AutomationTaskDto::taskId)
        val localByTaskId = localTasks.associateBy(AutomationTaskDto::taskId)
        if (serverByTaskId.keys != localByTaskId.keys) return true
        for ((taskId, serverTask) in serverByTaskId) {
            val localTask = localByTaskId.getValue(taskId)
            if (serverTask.roomName != localTask.roomName) return true
            if (serverTask.seatNumber != localTask.seatNumber) return true
            if (serverTask.enabled != localTask.enabled) return true
            if (serverTask.customWindows != localTask.customWindows) return true
        }
        return false
    }
}

/**
 * 候选条目类型（与 Window 端 `kind ∈ {'add','replace','remove'}` 等价）。
 *
 * 使用密封类而非 enum 是为了方便后续扩展携带类型相关字段（例如 remove 携带额外的
 * 「上次同步时间」），同时配合 Kotlin 编译期穷尽匹配给 SyncApplier 编写更强的契约。
 */
sealed class SyncKind {
    /** 服务端有、本地无：把受管字段写入本地。default_checked = true。 */
    data object Add : SyncKind()

    /** 两端匹配但受管字段不一致：用服务端值覆盖本地受管字段。default_checked = true。 */
    data object Replace : SyncKind()

    /**
     * 本地标记「受服务端管理」但服务端已无该 student_id：仅取消「受服务端管理」标记，
     * **不**物理删除 Room 行（Requirement 13-Q2）。default_checked = false。
     */
    data object Remove : SyncKind()
}

/**
 * 同步候选条目。Sync_Coverage_Confirmation 弹窗以此渲染列表。
 *
 * 字段语义与 design.md「Sync_Coverage_Confirmation 数据结构与字段口径」一致：
 * - `add` / `replace`：[serverPayload] 必填，[SyncKind.Add] 时 [localSummary] 为 null，
 *   [SyncKind.Replace] 时 [localSummary] 必填便于 UI 展示「替换前后对比」。
 * - `remove`：[serverPayload] 为 null，[localSummary] 必填便于 UI 展示要被取消管理的本地
 *   账号摘要。
 * - [defaultChecked] 给 UI 在初始化 Sync_Selection 时使用：add/replace 默认 true，
 *   remove 默认 false。
 */
data class SyncCandidate(
    val kind: SyncKind,
    val studentId: String,
    val serverPayload: ServerAccountSnapshot?,
    val localSummary: LocalAccountSummary?,
    val defaultChecked: Boolean,
)

/**
 * 服务端快照中的单个活跃账号，受管字段全集。
 *
 * 由 [AccountPoolSyncRepository.refreshActiveList] + 多次
 * [AccountPoolSyncRepository.getActiveAccountDetail] 合并产生：接口 A 提供清单，接口 B 提供
 * 明文密码与 `automation_tasks`。
 *
 * **注意**：[password] 是明文密码，**不**得持久化。SyncPlanner 仅做内存计算，候选条目交给
 * SyncApplier 时由 ViewModel 在该单次 Manual_Sync_Action 流程内引用，流程结束立即释放。
 *
 * [accountId] 任务 12.14 新增：生产 [LocalAccountStore] 实现把账号写入 Room
 * `active_accounts` 表时需要主键 `accountId`，由接口 A / B 响应直接透传。默认值 0 仅用于
 * 单元测试构造便利（不会落库）。
 */
data class ServerAccountSnapshot(
    val studentId: String,
    val password: String,
    val displayName: String,
    val automationTasks: List<AutomationTaskDto>,
    val accountId: Long = 0L,
)

/**
 * 本地存储中的单个账号视图（受管字段 + 管理标记）。
 *
 * 由调用方从 Room 表（账号表 + 自动任务表）合并装配；SyncPlanner 仅做纯函数计算，**不**
 * 直接读取 Room、**不**触发任何 IO。
 *
 * - [isServerManaged]：标记该账号是否「受服务端管理活跃账号」。该标记决定 remove 候选条目
 *   是否产出（design.md 伪代码：`if local_row.is_server_managed and sid not in server_by_sid`）。
 *   未标记的本地账号即使服务端清单缺失，也不会被列入 remove 候选，避免误清理用户手动添加
 *   的账号。
 * - [password]：本地保留的密码字段（如果客户端从未通过接口 B 拉取过则可为空字符串）。
 *   与服务端不一致即触发 replace。
 */
data class LocalAccount(
    val studentId: String,
    val password: String,
    val displayName: String,
    val automationTasks: List<AutomationTaskDto>,
    val isServerManaged: Boolean,
)

/**
 * 本地账号摘要，供 Sync_Coverage_Confirmation 弹窗在 replace / remove 候选条目上展示
 * 「替换前 / 移除前」的本地状态对比。
 *
 * 摘要字段集合与受管字段一致，但**不**回传明文密码（弹窗在 UI 层只展示遮罩字符串，避免
 * 屏幕录像 / 截图泄露）。
 */
data class LocalAccountSummary(
    val studentId: String,
    val displayName: String,
    val automationTaskCount: Int,
    val customWindowCount: Int,
)

/**
 * [LocalAccount] → [LocalAccountSummary]：摘要不携带明文密码，自动任务与自定义窗口仅以
 * 计数形式展示，避免在弹窗 UI 层暴露过多内部细节。
 */
internal fun LocalAccount.toSummary(): LocalAccountSummary =
    LocalAccountSummary(
        studentId = studentId,
        displayName = displayName,
        automationTaskCount = automationTasks.size,
        customWindowCount = automationTasks.sumOf { task -> task.customWindows.size },
    )
